# Project

In this Project, you will bring together many of the tools and techniques that you have learned throughout this course into a final project. You can choose from many different paths to get to the solution. 

### Business scenario

You work for a training organization that recently developed an introductory course about machine learning (ML). The course includes more than 40 videos that cover a broad range of ML topics. You have been asked to create an application that will students can use to quickly locate and view video content by searching for topics and key phrases.

You have downloaded all of the videos to an Amazon Simple Storage Service (Amazon S3) bucket. Your assignment is to produce a dashboard that meets your supervisor’s requirements.

## Project steps

To complete this project, you will follow these steps:

1. [Viewing the video files](#1.-Viewing-the-video-files)
2. [Transcribing the videos](#2.-Transcribing-the-videos)
3. [Normalizing the text](#3.-Normalizing-the-text)
4. [Extracting key phrases and topics](#4.-Extracting-key-phrases-and-topics)
5. [Creating the dashboard](#5.-Creating-the-dashboard)

## Useful information

The following cell contains some information that might be useful as you complete this project.

In [1]:
bucket = "c56161a939430l3396553t1w744137092661-labbucket-rn642jaq01e9"
job_data_access_role = 'arn:aws:iam::744137092661:role/service-role/c56161a939430l3396553t1w7-ComprehendDataAccessRole-1P24MSS91ADHP'

## 1. Viewing the video files
([Go to top](#Capstone-8:-Bringing-It-All-Together))


The source video files are located in the following shared Amazon Simple Storage Service (Amazon S3) bucket.

In [1]:
!aws s3 ls s3://aws-tc-largeobjects/CUR-TF-200-ACMNLP-1/video/

2021-04-26 16:17:33  410925369 Mod01_Course Overview.mp4
2021-04-26 16:10:02   39576695 Mod02_Intro.mp4
2021-04-26 16:31:23  302994828 Mod02_Sect01.mp4
2021-04-26 16:17:33  416563881 Mod02_Sect02.mp4
2021-04-26 16:17:33  318685583 Mod02_Sect03.mp4
2021-04-26 16:17:33  255877251 Mod02_Sect04.mp4
2021-04-26 16:23:51   99988046 Mod02_Sect05.mp4
2021-04-26 16:24:54   50700224 Mod02_WrapUp.mp4
2021-04-26 16:26:27   60627667 Mod03_Intro.mp4
2021-04-26 16:26:28  272229844 Mod03_Sect01.mp4
2021-04-26 16:27:06  309127124 Mod03_Sect02_part1.mp4
2021-04-26 16:27:06  195635527 Mod03_Sect02_part2.mp4
2021-04-26 16:28:03  123924818 Mod03_Sect02_part3.mp4
2021-04-26 16:31:28  171681915 Mod03_Sect03_part1.mp4
2021-04-26 16:32:07  285200083 Mod03_Sect03_part2.mp4
2021-04-26 16:33:17  105470345 Mod03_Sect03_part3.mp4
2021-04-26 16:35:10  157185651 Mod03_Sect04_part1.mp4
2021-04-26 16:36:27  187435635 Mod03_Sect04_part2.mp4
2021-04-26 16:36:40  280720369 Mod03_Sect04_part3.mp4
2021-04-26 16:40:01  443479

# DOWNLOADING VIDEOS USING AWS SECURITY CREDIENTIALS IN COMMAND PROMPT AND STORING THEM IN LOCAL DIRECTORY


Microsoft Windows [Version 10.0.22000.1696]
(c) Microsoft Corporation. All rights reserved.

C:\Users\messi>aws --version
aws-cli/2.11.13 Python/3.11.3 Windows/10 exe/AMD64 prompt/off

C:\Users\messi>aws configure
AWS Access Key ID [****************HVUH]: AKIA5A6M2Y6ZN3CKXXD4
AWS Secret Access Key [****************ye+y]: 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
Default region name [us-east-2]: us-east-2
Default output format [json]: json

C:\Users\messi>aws s3 ls s3://aws-tc-largeobjects/CUR-TF-200-ACMNLP-1/video/
2021-04-26 16:17:33  410925369 Mod01_Course Overview.mp4
2021-04-26 16:10:02   39576695 Mod02_Intro.mp4
2021-04-26 16:31:23  302994828 Mod02_Sect01.mp4
2021-04-26 16:17:33  416563881 Mod02_Sect02.mp4
2021-04-26 16:17:33  318685583 Mod02_Sect03.mp4
2021-04-26 16:17:33  255877251 Mod02_Sect04.mp4
2021-04-26 16:23:51   99988046 Mod02_Sect05.mp4
2021-04-26 16:24:54   50700224 Mod02_WrapUp.mp4
2021-04-26 16:26:27   60627667 Mod03_Intro.mp4
2021-04-26 16:26:28  272229844 Mod03_Sect01.mp4
2021-04-26 16:27:06  309127124 Mod03_Sect02_part1.mp4
2021-04-26 16:27:06  195635527 Mod03_Sect02_part2.mp4
2021-04-26 16:28:03  123924818 Mod03_Sect02_part3.mp4
2021-04-26 16:31:28  171681915 Mod03_Sect03_part1.mp4
2021-04-26 16:32:07  285200083 Mod03_Sect03_part2.mp4
2021-04-26 16:33:17  105470345 Mod03_Sect03_part3.mp4
2021-04-26 16:35:10  157185651 Mod03_Sect04_part1.mp4
2021-04-26 16:36:27  187435635 Mod03_Sect04_part2.mp4
2021-04-26 16:36:40  280720369 Mod03_Sect04_part3.mp4
2021-04-26 16:40:01  443479313 Mod03_Sect05.mp4
2021-04-26 16:40:08  234182186 Mod03_Sect06.mp4
2021-04-26 16:40:33  207718047 Mod03_Sect07_part1.mp4
2021-04-26 16:42:07  125592110 Mod03_Sect07_part2.mp4
2021-04-26 16:45:10  508500301 Mod03_Sect07_part3.mp4
2021-04-26 16:46:16  320126756 Mod03_Sect08.mp4
2021-04-26 16:46:43   41839508 Mod03_WrapUp.mp4
2021-04-26 16:46:55   34148489 Mod04_Intro.mp4
2021-04-26 16:48:24   84959465 Mod04_Sect01.mp4
2021-04-26 16:48:25  345182970 Mod04_Sect02_part1.mp4
2021-04-26 16:51:34  218661651 Mod04_Sect02_part2.mp4
2021-04-26 16:53:32  430140637 Mod04_Sect02_part3.mp4
2021-04-26 16:56:03   22036605 Mod04_WrapUp.mp4
2021-04-26 16:57:18   49187118 Mod05_Intro.mp4
2021-04-26 16:58:19  245798071 Mod05_Sect01_ver2.mp4
2021-04-26 16:58:50  233314835 Mod05_Sect02_part1_ver2.mp4
2021-04-26 16:59:14  348545306 Mod05_Sect02_part2.mp4
2021-04-26 16:59:17  239142711 Mod05_Sect03_part1.mp4
2021-04-26 17:06:04  267533559 Mod05_Sect03_part2.mp4
2021-04-26 17:06:06  212502220 Mod05_Sect03_part3.mp4
2021-04-26 17:06:48  206317022 Mod05_Sect03_part4_ver2.mp4
2021-04-26 17:06:48   60361230 Mod05_WrapUp_ver2.mp4
2021-04-26 17:09:14   35397860 Mod06_Intro.mp4
2021-04-26 17:09:24  845633599 Mod06_Sect01.mp4
2021-04-26 17:10:47  326126684 Mod06_Sect02.mp4
2021-04-26 17:12:26   19790740 Mod06_WrapUp.mp4
2021-04-26 17:12:56  131249036 Mod07_Sect01.mp4

C:\Users\messi>aws s3 sync s3://aws-tc-largeobjects/CUR-TF-200-ACMNLP-1/video/ "C:\Users\messi\Downloads\NLP"/ --exclude "*" --include "*.mp4"

## 2. Transcribing the videos
 ([Go to top](#Capstone-8:-Bringing-It-All-Together))

Use this section to implement your solution to transcribe the videos. 

In [1]:
import os
import math
import speech_recognition as sr
from moviepy.video.io.VideoFileClip import VideoFileClip

# set up file paths
path = r"C:/Users/messi/Downloads/NLP"
output_path = os.path.join(path, "Transcriptions")
os.makedirs(output_path, exist_ok=True)

# function to transcribe video segments
def transcribe_video(video_path, output_file, segment_duration=60, language='en-US'):
    # create audio file from video
    audio_file = video_path[:-4] + ".wav"
    try:
        video = VideoFileClip(video_path)
        audio = video.audio
        audio.write_audiofile(audio_file)
    except:
        print(f"Failed to create audio file for {video_path}")
        return

    # transcribe audio and save to text file
    r = sr.Recognizer()
    transcription = ""
    with sr.AudioFile(audio_file) as source:
        duration = source.DURATION
        total_duration = math.ceil(duration/segment_duration)
        for i in range(total_duration):
            audio = r.record(source, offset=i*segment_duration, duration=segment_duration)
            try:
                text = r.recognize_google(audio, language=language)
                transcription += text + " "
            except sr.UnknownValueError:
                print(f"Unable to transcribe segment {i+1} of {total_duration} for {video_path}")
                continue

    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(transcription)

    os.remove(audio_file)
    print(f"Transcription saved to: {output_file}")

# transcribe all videos in directory
for filename in os.listdir(path):
    if filename.endswith(".mp4"):
        print(f"Transcribing: {filename}")
        video_path = os.path.join(path, filename)
        output_filename = filename[:-4] + ".txt"
        output_file = os.path.join(output_path, output_filename)
        transcribe_video(video_path, output_file, segment_duration=60, language='en-US')


Transcribing: Mod01_Course Overview.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod01_Course Overview.wav


MoviePy - Done.
Unable to transcribe segment 5 of 11 for C:/Users/messi/Downloads/NLP\Mod01_Course Overview.mp4
Unable to transcribe segment 6 of 11 for C:/Users/messi/Downloads/NLP\Mod01_Course Overview.mp4
Unable to transcribe segment 7 of 11 for C:/Users/messi/Downloads/NLP\Mod01_Course Overview.mp4
Unable to transcribe segment 8 of 11 for C:/Users/messi/Downloads/NLP\Mod01_Course Overview.mp4
Unable to transcribe segment 9 of 11 for C:/Users/messi/Downloads/NLP\Mod01_Course Overview.mp4
Unable to transcribe segment 10 of 11 for C:/Users/messi/Downloads/NLP\Mod01_Course Overview.mp4
Unable to transcribe segment 11 of 11 for C:/Users/messi/Downloads/NLP\Mod01_Course Overview.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod01_Course Overview.txt
Transcribing: Mod02_Intro.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod02_Intro.wav


MoviePy - Done.
Unable to transcribe segment 2 of 2 for C:/Users/messi/Downloads/NLP\Mod02_Intro.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod02_Intro.txt
Transcribing: Mod02_Sect01.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod02_Sect01.wav


MoviePy - Done.
Unable to transcribe segment 4 of 8 for C:/Users/messi/Downloads/NLP\Mod02_Sect01.mp4
Unable to transcribe segment 5 of 8 for C:/Users/messi/Downloads/NLP\Mod02_Sect01.mp4
Unable to transcribe segment 6 of 8 for C:/Users/messi/Downloads/NLP\Mod02_Sect01.mp4
Unable to transcribe segment 7 of 8 for C:/Users/messi/Downloads/NLP\Mod02_Sect01.mp4
Unable to transcribe segment 8 of 8 for C:/Users/messi/Downloads/NLP\Mod02_Sect01.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod02_Sect01.txt
Transcribing: Mod02_Sect02.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod02_Sect02.wav


MoviePy - Done.
Unable to transcribe segment 5 of 11 for C:/Users/messi/Downloads/NLP\Mod02_Sect02.mp4
Unable to transcribe segment 6 of 11 for C:/Users/messi/Downloads/NLP\Mod02_Sect02.mp4
Unable to transcribe segment 7 of 11 for C:/Users/messi/Downloads/NLP\Mod02_Sect02.mp4
Unable to transcribe segment 8 of 11 for C:/Users/messi/Downloads/NLP\Mod02_Sect02.mp4
Unable to transcribe segment 9 of 11 for C:/Users/messi/Downloads/NLP\Mod02_Sect02.mp4
Unable to transcribe segment 10 of 11 for C:/Users/messi/Downloads/NLP\Mod02_Sect02.mp4
Unable to transcribe segment 11 of 11 for C:/Users/messi/Downloads/NLP\Mod02_Sect02.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod02_Sect02.txt
Transcribing: Mod02_Sect03.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod02_Sect03.wav


MoviePy - Done.
Unable to transcribe segment 4 of 9 for C:/Users/messi/Downloads/NLP\Mod02_Sect03.mp4
Unable to transcribe segment 5 of 9 for C:/Users/messi/Downloads/NLP\Mod02_Sect03.mp4
Unable to transcribe segment 6 of 9 for C:/Users/messi/Downloads/NLP\Mod02_Sect03.mp4
Unable to transcribe segment 7 of 9 for C:/Users/messi/Downloads/NLP\Mod02_Sect03.mp4
Unable to transcribe segment 8 of 9 for C:/Users/messi/Downloads/NLP\Mod02_Sect03.mp4
Unable to transcribe segment 9 of 9 for C:/Users/messi/Downloads/NLP\Mod02_Sect03.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod02_Sect03.txt
Transcribing: Mod02_Sect04.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod02_Sect04.wav


MoviePy - Done.
Unable to transcribe segment 4 of 7 for C:/Users/messi/Downloads/NLP\Mod02_Sect04.mp4
Unable to transcribe segment 5 of 7 for C:/Users/messi/Downloads/NLP\Mod02_Sect04.mp4
Unable to transcribe segment 6 of 7 for C:/Users/messi/Downloads/NLP\Mod02_Sect04.mp4
Unable to transcribe segment 7 of 7 for C:/Users/messi/Downloads/NLP\Mod02_Sect04.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod02_Sect04.txt
Transcribing: Mod02_Sect05.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod02_Sect05.wav


MoviePy - Done.
Unable to transcribe segment 3 of 3 for C:/Users/messi/Downloads/NLP\Mod02_Sect05.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod02_Sect05.txt
Transcribing: Mod02_WrapUp.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod02_WrapUp.wav


MoviePy - Done.
Unable to transcribe segment 2 of 2 for C:/Users/messi/Downloads/NLP\Mod02_WrapUp.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod02_WrapUp.txt
Transcribing: Mod03_Intro.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Intro.wav


MoviePy - Done.
Unable to transcribe segment 2 of 2 for C:/Users/messi/Downloads/NLP\Mod03_Intro.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Intro.txt
Transcribing: Mod03_Sect01.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect01.wav


MoviePy - Done.
Unable to transcribe segment 4 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect01.mp4
Unable to transcribe segment 5 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect01.mp4
Unable to transcribe segment 6 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect01.mp4
Unable to transcribe segment 7 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect01.mp4
Unable to transcribe segment 8 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect01.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect01.txt
Transcribing: Mod03_Sect02_part1.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect02_part1.wav


MoviePy - Done.
Unable to transcribe segment 4 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect02_part1.mp4
Unable to transcribe segment 5 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect02_part1.mp4
Unable to transcribe segment 6 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect02_part1.mp4
Unable to transcribe segment 7 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect02_part1.mp4
Unable to transcribe segment 8 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect02_part1.mp4
Unable to transcribe segment 9 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect02_part1.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect02_part1.txt
Transcribing: Mod03_Sect02_part2.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect02_part2.wav


MoviePy - Done.
Unable to transcribe segment 4 of 6 for C:/Users/messi/Downloads/NLP\Mod03_Sect02_part2.mp4
Unable to transcribe segment 5 of 6 for C:/Users/messi/Downloads/NLP\Mod03_Sect02_part2.mp4
Unable to transcribe segment 6 of 6 for C:/Users/messi/Downloads/NLP\Mod03_Sect02_part2.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect02_part2.txt
Transcribing: Mod03_Sect02_part3.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect02_part3.wav


MoviePy - Done.
Unable to transcribe segment 3 of 4 for C:/Users/messi/Downloads/NLP\Mod03_Sect02_part3.mp4
Unable to transcribe segment 4 of 4 for C:/Users/messi/Downloads/NLP\Mod03_Sect02_part3.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect02_part3.txt
Transcribing: Mod03_Sect03_part1.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect03_part1.wav


MoviePy - Done.
Unable to transcribe segment 3 of 5 for C:/Users/messi/Downloads/NLP\Mod03_Sect03_part1.mp4
Unable to transcribe segment 4 of 5 for C:/Users/messi/Downloads/NLP\Mod03_Sect03_part1.mp4
Unable to transcribe segment 5 of 5 for C:/Users/messi/Downloads/NLP\Mod03_Sect03_part1.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect03_part1.txt
Transcribing: Mod03_Sect03_part2.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect03_part2.wav


MoviePy - Done.
Unable to transcribe segment 4 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect03_part2.mp4
Unable to transcribe segment 5 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect03_part2.mp4
Unable to transcribe segment 6 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect03_part2.mp4
Unable to transcribe segment 7 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect03_part2.mp4
Unable to transcribe segment 8 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect03_part2.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect03_part2.txt
Transcribing: Mod03_Sect03_part3.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect03_part3.wav


MoviePy - Done.
Unable to transcribe segment 3 of 3 for C:/Users/messi/Downloads/NLP\Mod03_Sect03_part3.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect03_part3.txt
Transcribing: Mod03_Sect04_part1.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect04_part1.wav


MoviePy - Done.
Unable to transcribe segment 3 of 5 for C:/Users/messi/Downloads/NLP\Mod03_Sect04_part1.mp4
Unable to transcribe segment 4 of 5 for C:/Users/messi/Downloads/NLP\Mod03_Sect04_part1.mp4
Unable to transcribe segment 5 of 5 for C:/Users/messi/Downloads/NLP\Mod03_Sect04_part1.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect04_part1.txt
Transcribing: Mod03_Sect04_part2.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect04_part2.wav


MoviePy - Done.
Unable to transcribe segment 3 of 5 for C:/Users/messi/Downloads/NLP\Mod03_Sect04_part2.mp4
Unable to transcribe segment 4 of 5 for C:/Users/messi/Downloads/NLP\Mod03_Sect04_part2.mp4
Unable to transcribe segment 5 of 5 for C:/Users/messi/Downloads/NLP\Mod03_Sect04_part2.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect04_part2.txt
Transcribing: Mod03_Sect04_part3.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect04_part3.wav


MoviePy - Done.
Unable to transcribe segment 4 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect04_part3.mp4
Unable to transcribe segment 5 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect04_part3.mp4
Unable to transcribe segment 6 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect04_part3.mp4
Unable to transcribe segment 7 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect04_part3.mp4
Unable to transcribe segment 8 of 8 for C:/Users/messi/Downloads/NLP\Mod03_Sect04_part3.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect04_part3.txt
Transcribing: Mod03_Sect05.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect05.wav


MoviePy - Done.
Unable to transcribe segment 5 of 12 for C:/Users/messi/Downloads/NLP\Mod03_Sect05.mp4
Unable to transcribe segment 6 of 12 for C:/Users/messi/Downloads/NLP\Mod03_Sect05.mp4
Unable to transcribe segment 7 of 12 for C:/Users/messi/Downloads/NLP\Mod03_Sect05.mp4
Unable to transcribe segment 8 of 12 for C:/Users/messi/Downloads/NLP\Mod03_Sect05.mp4
Unable to transcribe segment 9 of 12 for C:/Users/messi/Downloads/NLP\Mod03_Sect05.mp4
Unable to transcribe segment 10 of 12 for C:/Users/messi/Downloads/NLP\Mod03_Sect05.mp4
Unable to transcribe segment 11 of 12 for C:/Users/messi/Downloads/NLP\Mod03_Sect05.mp4
Unable to transcribe segment 12 of 12 for C:/Users/messi/Downloads/NLP\Mod03_Sect05.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect05.txt
Transcribing: Mod03_Sect06.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect06.wav


MoviePy - Done.
Unable to transcribe segment 4 of 7 for C:/Users/messi/Downloads/NLP\Mod03_Sect06.mp4
Unable to transcribe segment 5 of 7 for C:/Users/messi/Downloads/NLP\Mod03_Sect06.mp4
Unable to transcribe segment 6 of 7 for C:/Users/messi/Downloads/NLP\Mod03_Sect06.mp4
Unable to transcribe segment 7 of 7 for C:/Users/messi/Downloads/NLP\Mod03_Sect06.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect06.txt
Transcribing: Mod03_Sect07_part1.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect07_part1.wav


MoviePy - Done.
Unable to transcribe segment 3 of 4 for C:/Users/messi/Downloads/NLP\Mod03_Sect07_part1.mp4
Unable to transcribe segment 4 of 4 for C:/Users/messi/Downloads/NLP\Mod03_Sect07_part1.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect07_part1.txt
Transcribing: Mod03_Sect07_part2.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect07_part2.wav


MoviePy - Done.
Unable to transcribe segment 3 of 4 for C:/Users/messi/Downloads/NLP\Mod03_Sect07_part2.mp4
Unable to transcribe segment 4 of 4 for C:/Users/messi/Downloads/NLP\Mod03_Sect07_part2.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect07_part2.txt
Transcribing: Mod03_Sect07_part3.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect07_part3.wav


MoviePy - Done.
Unable to transcribe segment 4 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect07_part3.mp4
Unable to transcribe segment 5 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect07_part3.mp4
Unable to transcribe segment 6 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect07_part3.mp4
Unable to transcribe segment 7 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect07_part3.mp4
Unable to transcribe segment 8 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect07_part3.mp4
Unable to transcribe segment 9 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect07_part3.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect07_part3.txt
Transcribing: Mod03_Sect08.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_Sect08.wav


MoviePy - Done.
Unable to transcribe segment 4 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect08.mp4
Unable to transcribe segment 5 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect08.mp4
Unable to transcribe segment 6 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect08.mp4
Unable to transcribe segment 7 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect08.mp4
Unable to transcribe segment 8 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect08.mp4
Unable to transcribe segment 9 of 9 for C:/Users/messi/Downloads/NLP\Mod03_Sect08.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_Sect08.txt
Transcribing: Mod03_WrapUp.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod03_WrapUp.wav


MoviePy - Done.
Unable to transcribe segment 2 of 2 for C:/Users/messi/Downloads/NLP\Mod03_WrapUp.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod03_WrapUp.txt
Transcribing: Mod04_Intro.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod04_Intro.wav


MoviePy - Done.
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod04_Intro.txt
Transcribing: Mod04_Sect01.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod04_Sect01.wav


MoviePy - Done.
Unable to transcribe segment 3 of 3 for C:/Users/messi/Downloads/NLP\Mod04_Sect01.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod04_Sect01.txt
Transcribing: Mod04_Sect02_part1.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod04_Sect02_part1.wav


MoviePy - Done.
Unable to transcribe segment 4 of 9 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part1.mp4
Unable to transcribe segment 5 of 9 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part1.mp4
Unable to transcribe segment 6 of 9 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part1.mp4
Unable to transcribe segment 7 of 9 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part1.mp4
Unable to transcribe segment 8 of 9 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part1.mp4
Unable to transcribe segment 9 of 9 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part1.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod04_Sect02_part1.txt
Transcribing: Mod04_Sect02_part2.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod04_Sect02_part2.wav


MoviePy - Done.
Unable to transcribe segment 4 of 6 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part2.mp4
Unable to transcribe segment 5 of 6 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part2.mp4
Unable to transcribe segment 6 of 6 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part2.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod04_Sect02_part2.txt
Transcribing: Mod04_Sect02_part3.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod04_Sect02_part3.wav


MoviePy - Done.
Unable to transcribe segment 5 of 12 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part3.mp4
Unable to transcribe segment 6 of 12 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part3.mp4
Unable to transcribe segment 7 of 12 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part3.mp4
Unable to transcribe segment 8 of 12 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part3.mp4
Unable to transcribe segment 9 of 12 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part3.mp4
Unable to transcribe segment 10 of 12 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part3.mp4
Unable to transcribe segment 11 of 12 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part3.mp4
Unable to transcribe segment 12 of 12 for C:/Users/messi/Downloads/NLP\Mod04_Sect02_part3.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod04_Sect02_part3.txt
Transcribing: Mod04_WrapUp.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod04_WrapUp.wav


MoviePy - Done.
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod04_WrapUp.txt
Transcribing: Mod05_Intro.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod05_Intro.wav


MoviePy - Done.
Unable to transcribe segment 2 of 2 for C:/Users/messi/Downloads/NLP\Mod05_Intro.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod05_Intro.txt
Transcribing: Mod05_Sect01_ver2.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod05_Sect01_ver2.wav


MoviePy - Done.
Unable to transcribe segment 4 of 8 for C:/Users/messi/Downloads/NLP\Mod05_Sect01_ver2.mp4
Unable to transcribe segment 5 of 8 for C:/Users/messi/Downloads/NLP\Mod05_Sect01_ver2.mp4
Unable to transcribe segment 6 of 8 for C:/Users/messi/Downloads/NLP\Mod05_Sect01_ver2.mp4
Unable to transcribe segment 7 of 8 for C:/Users/messi/Downloads/NLP\Mod05_Sect01_ver2.mp4
Unable to transcribe segment 8 of 8 for C:/Users/messi/Downloads/NLP\Mod05_Sect01_ver2.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod05_Sect01_ver2.txt
Transcribing: Mod05_Sect02_part1_ver2.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod05_Sect02_part1_ver2.wav


MoviePy - Done.
Unable to transcribe segment 4 of 9 for C:/Users/messi/Downloads/NLP\Mod05_Sect02_part1_ver2.mp4
Unable to transcribe segment 5 of 9 for C:/Users/messi/Downloads/NLP\Mod05_Sect02_part1_ver2.mp4
Unable to transcribe segment 6 of 9 for C:/Users/messi/Downloads/NLP\Mod05_Sect02_part1_ver2.mp4
Unable to transcribe segment 7 of 9 for C:/Users/messi/Downloads/NLP\Mod05_Sect02_part1_ver2.mp4
Unable to transcribe segment 8 of 9 for C:/Users/messi/Downloads/NLP\Mod05_Sect02_part1_ver2.mp4
Unable to transcribe segment 9 of 9 for C:/Users/messi/Downloads/NLP\Mod05_Sect02_part1_ver2.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod05_Sect02_part1_ver2.txt
Transcribing: Mod05_Sect02_part2.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod05_Sect02_part2.wav


MoviePy - Done.
Unable to transcribe segment 5 of 10 for C:/Users/messi/Downloads/NLP\Mod05_Sect02_part2.mp4
Unable to transcribe segment 6 of 10 for C:/Users/messi/Downloads/NLP\Mod05_Sect02_part2.mp4
Unable to transcribe segment 7 of 10 for C:/Users/messi/Downloads/NLP\Mod05_Sect02_part2.mp4
Unable to transcribe segment 8 of 10 for C:/Users/messi/Downloads/NLP\Mod05_Sect02_part2.mp4
Unable to transcribe segment 9 of 10 for C:/Users/messi/Downloads/NLP\Mod05_Sect02_part2.mp4
Unable to transcribe segment 10 of 10 for C:/Users/messi/Downloads/NLP\Mod05_Sect02_part2.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod05_Sect02_part2.txt
Transcribing: Mod05_Sect03_part1.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod05_Sect03_part1.wav


MoviePy - Done.
Unable to transcribe segment 4 of 7 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part1.mp4
Unable to transcribe segment 5 of 7 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part1.mp4
Unable to transcribe segment 6 of 7 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part1.mp4
Unable to transcribe segment 7 of 7 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part1.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod05_Sect03_part1.txt
Transcribing: Mod05_Sect03_part2.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod05_Sect03_part2.wav


MoviePy - Done.
Unable to transcribe segment 4 of 7 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part2.mp4
Unable to transcribe segment 5 of 7 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part2.mp4
Unable to transcribe segment 6 of 7 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part2.mp4
Unable to transcribe segment 7 of 7 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part2.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod05_Sect03_part2.txt
Transcribing: Mod05_Sect03_part3.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod05_Sect03_part3.wav


MoviePy - Done.
Unable to transcribe segment 4 of 6 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part3.mp4
Unable to transcribe segment 5 of 6 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part3.mp4
Unable to transcribe segment 6 of 6 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part3.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod05_Sect03_part3.txt
Transcribing: Mod05_Sect03_part4_ver2.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod05_Sect03_part4_ver2.wav


MoviePy - Done.
Unable to transcribe segment 4 of 7 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part4_ver2.mp4
Unable to transcribe segment 5 of 7 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part4_ver2.mp4
Unable to transcribe segment 6 of 7 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part4_ver2.mp4
Unable to transcribe segment 7 of 7 for C:/Users/messi/Downloads/NLP\Mod05_Sect03_part4_ver2.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod05_Sect03_part4_ver2.txt
Transcribing: Mod05_WrapUp_ver2.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod05_WrapUp_ver2.wav


MoviePy - Done.
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod05_WrapUp_ver2.txt
Transcribing: Mod06_Intro.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod06_Intro.wav


MoviePy - Done.
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod06_Intro.txt
Transcribing: Mod06_Sect01.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod06_Sect01.wav


MoviePy - Done.
Unable to transcribe segment 5 of 12 for C:/Users/messi/Downloads/NLP\Mod06_Sect01.mp4
Unable to transcribe segment 6 of 12 for C:/Users/messi/Downloads/NLP\Mod06_Sect01.mp4
Unable to transcribe segment 7 of 12 for C:/Users/messi/Downloads/NLP\Mod06_Sect01.mp4
Unable to transcribe segment 8 of 12 for C:/Users/messi/Downloads/NLP\Mod06_Sect01.mp4
Unable to transcribe segment 9 of 12 for C:/Users/messi/Downloads/NLP\Mod06_Sect01.mp4
Unable to transcribe segment 10 of 12 for C:/Users/messi/Downloads/NLP\Mod06_Sect01.mp4
Unable to transcribe segment 11 of 12 for C:/Users/messi/Downloads/NLP\Mod06_Sect01.mp4
Unable to transcribe segment 12 of 12 for C:/Users/messi/Downloads/NLP\Mod06_Sect01.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod06_Sect01.txt
Transcribing: Mod06_Sect02.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod06_Sect02.wav


MoviePy - Done.
Unable to transcribe segment 4 of 9 for C:/Users/messi/Downloads/NLP\Mod06_Sect02.mp4
Unable to transcribe segment 5 of 9 for C:/Users/messi/Downloads/NLP\Mod06_Sect02.mp4
Unable to transcribe segment 6 of 9 for C:/Users/messi/Downloads/NLP\Mod06_Sect02.mp4
Unable to transcribe segment 7 of 9 for C:/Users/messi/Downloads/NLP\Mod06_Sect02.mp4
Unable to transcribe segment 8 of 9 for C:/Users/messi/Downloads/NLP\Mod06_Sect02.mp4
Unable to transcribe segment 9 of 9 for C:/Users/messi/Downloads/NLP\Mod06_Sect02.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod06_Sect02.txt
Transcribing: Mod06_WrapUp.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod06_WrapUp.wav


MoviePy - Done.
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod06_WrapUp.txt
Transcribing: Mod07_Sect01.mp4
MoviePy - Writing audio in C:/Users/messi/Downloads/NLP\Mod07_Sect01.wav


MoviePy - Done.
Unable to transcribe segment 3 of 4 for C:/Users/messi/Downloads/NLP\Mod07_Sect01.mp4
Unable to transcribe segment 4 of 4 for C:/Users/messi/Downloads/NLP\Mod07_Sect01.mp4
Transcription saved to: C:/Users/messi/Downloads/NLP\Transcriptions\Mod07_Sect01.txt


## 3. Normalizing the text
([Go to top](#Capstone-8:-Bringing-It-All-Together))

Use this section to perform any text normalization steps that are necessary for your solution.

In [2]:
pip install SpeechRecognition

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [3]:
pip install moviepy

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [4]:
pip install nltk

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [55]:
import os
import re
import string
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

# set up file paths
path = r"C:/Users/messi/Downloads/NLP"
transcription_path = os.path.join(path, "Transcriptions")
normalized_path = os.path.join(path, "Normalized")
os.makedirs(normalized_path, exist_ok=True)

# define function to normalize text
def normalize_text(text):
    # convert to lowercase
    text = text.lower()

    # remove punctuation
    text = text.translate(str.maketrans('', '', string.punctuation))

    # remove numbers
    text = re.sub(r'\d+', '', text)

    # remove stopwords
    stop_words = set(stopwords.words('english'))
    word_tokens = word_tokenize(text)
    filtered_text = [word for word in word_tokens if word not in stop_words]
    text = ' '.join(filtered_text)

    return text

# normalize all transcriptions in directory
for filename in os.listdir(transcription_path):
    if filename.endswith(".txt"):
        print(f"Normalizing: {filename}")
        input_file = os.path.join(transcription_path, filename)
        output_file = os.path.join(normalized_path, filename)

        # read input file
        with open(input_file, 'r', encoding='utf-8') as f:
            text = f.read()

        # normalize text
        normalized_text = normalize_text(text)

        # write output file
        with open(output_file, 'w', encoding='utf-8') as f:
            f.write(normalized_text)

        print(f"Normalized text saved to: {output_file}")


Normalizing: Mod01_Course Overview.txt
Normalized text saved to: C:/Users/messi/Downloads/NLP\Normalized\Mod01_Course Overview.txt
Normalizing: Mod02_Intro.txt
Normalized text saved to: C:/Users/messi/Downloads/NLP\Normalized\Mod02_Intro.txt
Normalizing: Mod02_Sect01.txt
Normalized text saved to: C:/Users/messi/Downloads/NLP\Normalized\Mod02_Sect01.txt
Normalizing: Mod02_Sect02.txt
Normalized text saved to: C:/Users/messi/Downloads/NLP\Normalized\Mod02_Sect02.txt
Normalizing: Mod02_Sect03.txt
Normalized text saved to: C:/Users/messi/Downloads/NLP\Normalized\Mod02_Sect03.txt
Normalizing: Mod02_Sect04.txt
Normalized text saved to: C:/Users/messi/Downloads/NLP\Normalized\Mod02_Sect04.txt
Normalizing: Mod02_Sect05.txt
Normalized text saved to: C:/Users/messi/Downloads/NLP\Normalized\Mod02_Sect05.txt
Normalizing: Mod02_WrapUp.txt
Normalized text saved to: C:/Users/messi/Downloads/NLP\Normalized\Mod02_WrapUp.txt
Normalizing: Mod03_Intro.txt
Normalized text saved to: C:/Users/messi/Downloads/

## 4. Extracting key phrases and topics
([Go to top](#Capstone-8:-Bringing-It-All-Together))

Use this section to extract the key phrases and topics from the videos.

In [6]:
pip install textblob

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [7]:
pip install gensim

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [56]:
import os
import string
from textblob import TextBlob
from gensim import corpora, models
from gensim.models import CoherenceModel
from nltk.corpus import stopwords

directory = "C:/Users/messi/Downloads/NLP/Normalized"
output_directory = "C:/Users/messi/Downloads/NLP"
os.makedirs(output_directory, exist_ok=True)

def extract_phrases_topics(text):
    # preprocess text
    text = text.lower()
    text = text.translate(str.maketrans('', '', string.punctuation))
    words = TextBlob(text).words
    stop_words = stopwords.words('english')
    words = [word.lemma for word in words if len(word.lemma) > 2 and word.lemma not in stop_words]

    # extract key phrases
    phrases = TextBlob(text).noun_phrases

    # create corpus and dictionary for LDA model
    corpus = [words]
    dictionary = corpora.Dictionary(corpus)

    # create bag of words representation of corpus
    bow_corpus = [dictionary.doc2bow(doc) for doc in corpus]

    # train LDA model on corpus
    coherence_values = []
    for num_topics in range(2, 6):
        lda_model = models.LdaModel(bow_corpus, num_topics=num_topics, id2word=dictionary, passes=10)
        coherence_model_lda = CoherenceModel(model=lda_model, texts=corpus, dictionary=dictionary, coherence='c_v')
        coherence_values.append(coherence_model_lda.get_coherence())
    best_num_topics = coherence_values.index(max(coherence_values)) + 2
    lda_model = models.LdaModel(bow_corpus, num_topics=best_num_topics, id2word=dictionary, passes=10)

    # extract topics and associated keywords
    topics = lda_model.show_topics(num_topics=best_num_topics, num_words=5, formatted=False)

    return phrases, topics

# loop through files in directory
for filename in os.listdir(directory):
    # check that file is a text file
    if filename.endswith('.txt'):
        # construct full filepath
        filepath = os.path.join(directory, filename)
        
        # read file
        with open(filepath, 'r') as f:
            text = f.read()
            
        # extract key phrases and topics
        phrases, topics = extract_phrases_topics(text)
    
        # create output file
        output_filepath = os.path.join(output_directory, filename)
        with open(output_filepath, 'w') as f:
            f.write("Key Phrases:\n")
            for phrase in phrases:
                f.write(f"- {phrase}\n")
            
            f.write("\nTopics:\n")
            for topic in topics:
                words = [word[0] for word in topic[1]]
                f.write("- ")
                f.write(" ".join(words))
                f.write("\n")


## 5. Creating the dashboard
([Go to top](#Capstone-8:-Bringing-It-All-Together))

Use this section to create the dashboard for your solution.

In [1]:
import os
from IPython.display import display, clear_output, Video
import ipywidgets as widgets

base_directory = "C:/Users/messi/Downloads/NLP"
transcribed_directory = os.path.join(base_directory, "Transcription")
keyphrases_topics_directory = os.path.join(base_directory, "Key Phrases and Topics")

# List all video files
video_files = [f for f in os.listdir(base_directory) if f.endswith(".mp4")]

# Create widgets
video_selector = widgets.Select(options=video_files, description="Select video:")
display_button = widgets.Button(description="Play Selected Video")
video_player = widgets.Video(width=600, height=450)
transcribed_output = widgets.Output()
keyphrases_topics_output = widgets.Output()

def play_video():
    if video_player.value:
        video_player.unobserve_all()
        video_player.observe(lambda change: play_video(), names='playing')
        video_player.observe(lambda change: pause_video(), names='paused')
        display(video_player)

def pause_video():
    if video_player.value:
        video_player.unobserve_all()
        video_player.observe(lambda change: play_video(), names='playing')
        video_player.observe(lambda change: pause_video(), names='paused')
        clear_output(wait=True)
        with transcription_output:
            display(widgets.HTML("<h3>Paused</h3>"))
        display(video_player)

def display_info(button):
    clear_output(wait=True)
    display(video_selector)
    display(display_button)
    
    selected_video = video_selector.value
    video_name = os.path.splitext(selected_video)[0]
    
    # Play the video
    video_path = os.path.join(base_directory, selected_video)
    video_player.value = open(video_path, "rb").read()
    play_video()
    
    # Print transcribed text
    transcription_file = os.path.join(transcription_directory, f"{video_name}.txt")
    with open(transcription_file, "r") as f:
        transcribed_text = f.read()
    
    with transcription_output:
        clear_output(wait=True)
        display(widgets.HTML("<h3>Transcribed Text:</h3>"))
        display(widgets.HTML(f"<p>{transcribed_text}</p>"))
    
    # Print key phrases and topics
    keyphrases_topics_file = os.path.join(keyphrases_topics_directory, f"{video_name}.txt")
    with open(keyphrases_topics_file, "r") as f:
        keyphrases_topics = f.read()
    
    with keyphrases_topics_output:
        clear_output(wait=True)
        display(widgets.HTML("<h3>Key Phrases and Topics:</h3>"))
        display(widgets.HTML(f"<p>{keyphrases_topics}</p>"))

    # Show transcription and key phrases/topics
    display(transcription_output)
    display(keyphrases_topics_output)

display_button.on_click(display_info)

display(widgets.VBox([video_selector, display_button]))
display(widgets.VBox([video_player, transcription_output, keyphrases_topics_output]))


NameError: name 'transcription_output' is not defined